### PR TITLE
Code cleanup - eliminate build warnings

### PR DESCRIPTION
**Pull request by schlaman-ms**
The only purpose of this PR is to eliminate build warnings.
There are no new features or changes to logic or fixes.

## Changes
For two warnings that could not be eliminated, $(MSBuildWarningsAsMessages) was used.
For the Newtonsoft version conflicts, MSB3277.
For the copy files errors, MSB3030.
The copy file error occurs even though the files eventually do copy.

Here is an example of the MSBuildWarningsAsMessages
  <PropertyGroup>
 <MSBuildWarningsAsMessages>
     $(MSBuildWarningsAsMessages);MSB3277;MSB3030
 </MSBuildWarningsAsMessages>
  </PropertyGroup>

The rest of the fixes are like xml comments missing, and correcting order of static vs public, fields vs properties.

For significant contributions please make sure you have completed the following items:

* [ ] `ReleaseHistory.md` updated for non-trivial changes
* [ ] Added unit tests

### CODE REVIEW

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj [hunk 1]
```diff
@@ -11,6 +11,12 @@
     <PackageId>Sarif.PatternMatcher.AzureDevOpsConfiguration</PackageId>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <MSBuildWarningsAsMessages>
+	    $(MSBuildWarningsAsMessages);MSB3277;MSB3030
```
Comment (shaopeng-gh) [inline Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:16]: >MSBuildWarningsAsMessages [](http://example.com/codeflow?start=7&length=25)

Just a question if this is what we want, or fix the warning.

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs [hunk 2]
```diff
@@ -350,36 +373,14 @@ internal static void UpdateLevelKind(MatchExpression matchExpression)
             return result;
         }
 
-        public override AnalyzeContext InitializeGlobalContextFromOptions(AnalyzeOptions options, ref AnalyzeContext context)
```
Comment (shaopeng-gh) [inline Src/Sarif.PatternMatcher/AnalyzeCommand.cs:353]: >InitializeGlobalContextFromOptions [](http://example.com/codeflow?start=39&length=34)

It looks like much change here, compare to other small; changes, could you explain a bit the change here

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 1 above)
Comment (yongyan-gh) [inline Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:16]: >	     [](http://example.com/codeflow?start=0&length=5)

mixing of tab and space characters

Comment (yongyan-gh) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:16]: same to all other csproj files

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj [hunk 3]
```diff
@@ -11,6 +11,12 @@
     <PackageId>Sarif.PatternMatcher.AzureDevOpsConfiguration</PackageId>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <MSBuildWarningsAsMessages>
```
Comment (suvamM) [inline Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: @schlaman-ms Can you show the before and after due to this change? The build will now result in these messages? Or will they be suppressed entirely?

File: Src/Plugins/SalModernization/SalModernization.csproj [hunk 4]
```diff
@@ -11,6 +11,12 @@
     <PackageId>Sarif.PatternMatcher.SalModernization</PackageId>
   </PropertyGroup>
 
+  <PropertyGroup>
+	<MSBuildWarningsAsMessages>
+	    $(MSBuildWarningsAsMessages);MSB3277;MSB3030
+	</MSBuildWarningsAsMessages>
+  </PropertyGroup>
```
Comment (suvamM) [inline Src/Plugins/SalModernization/SalModernization.csproj:18]: @schlaman-ms There seems to be code duplication. Can we move this to build.common.props perhaps?

File: Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs [hunk 5]
```diff
@@ -39,6 +39,9 @@ public AlibabaEcsRequestSigner(DateTime? dateTime = null, string signatureNonce
         /// Sign the HTTP request using Alibaba Cloud access key and secret so it
         /// can be used to access Alibaba Cloud service REST API.
         /// </summary>
+        /// <param name="request">httpRequest.</param>
```
Comment (suvamM) [inline Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs:42]: >httpRequest. [](http://example.com/codeflow?start=34&length=12)

@schlaman-ms This isn't a great XML  comment. Can you add a bit more description?

File: Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs [hunk 6]
```diff
@@ -29,11 +29,11 @@ internal class SpamTestRule : Skimmer<AnalyzeContext>, IOptionsProvider
         internal static Random s_random = new Random(s_seed);
 
         [ThreadStatic]
-        internal static TestRuleBehaviors s_testRuleBehaviors;
+        private static TestRuleBehaviors s_testRuleBehaviors;
 
         public SpamTestRule()
         {
-            if (s_testRuleBehaviors.HasFlag(TestRuleBehaviors.RaiseExceptionInvokingConstructor))
+            if (S_testRuleBehaviors.HasFlag(TestRuleBehaviors.RaiseExceptionInvokingConstructor))
```
Comment (suvamM) [inline Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs:36]: >S_testRuleBehaviors [](http://example.com/codeflow?start=16&length=19)

@schlaman-ms I didn't fully follow this change. The variable name is still s_testRuleBehaviors right? Why capitalize the S?

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 2 above)
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:353]: There should be no functionality changes AT ALL. This is only to make the compiler happy to have things in the order it wants.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 3 above)
Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: No messages will show in the build at all with these suppressions.
Before you had many many lines of warning from the Newtonsoft conflict. Then there were the 4 copy warning lines.
Now there are zero lines generated as a result of these suppressions.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 1 above)
Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:16]: It is not possible to fix either of the sources of the warnings, therefore I need to suppress them.

Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:16]: I can't update newtonsoft to eliminate the conflict warning and I can't get rid of the copy commands that sometimes don't copy and give warning but eventually copy.

Comment (suvamM) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj]: @schlaman-ms Do we have an understanding on why this transient fault is occurring for making copies? I agree with @shaopeng-gh that we should try to fix it, rather than suppress it. Is this a known issue?

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 3 above)
Comment (suvamM) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: Not sure if this actually resolves the underlying problem.

File: Src/Plugins/SalModernization/SalModernization.csproj (hunk 4 above)
Comment (schlaman-ms) [reply Src/Plugins/SalModernization/SalModernization.csproj:18]: Well I intentionally only put the MSBuildWarningsAsMessages entries in these projects as the suppressions only occur in the project where the entry is.

If I put this in build.common.props then I would suppress those errors globally and that wouldn't be good.
There is already a certain amount of risk by suppressing that you silence something that shouldn't be silenced. If in a given project, I only need to silence one warning then that project only gets the one MSBxxxx entry.

File: Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs (hunk 5 above)
Comment (schlaman-ms) [reply Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs:42]: Yes, there were warnings generated because there were no XML parameter entries in the XML Comment.
I'll look at bit more at the code to come up with a better comment. Easier if it was my own code

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 2 above)
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:353]: Yes, sorry, this was a really messy one. properties not before fields. Constructor not before fields. public not before static.
I pretty much had to completely rearrange to satisfy all the different conditions.

File: Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs (hunk 6 above)
Comment (schlaman-ms) [reply Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs:36]: Well the issue here is to eliminate the warning I took the automatic visual studio change.
It does look like all that changed is internal to private and then s_test... to S_test...
But s_test... still exists and is the backing field and the new property S_test... is added.
That's why it looks hard to follow and odd why the case change s_ --> S_

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 1 above)
Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj]: Transient isn't the correct word here really.
It will always have the failure to copy error, but by the end of the build, in each case for the 4 different warnings, each of those 4 files ends up copied.

The copy parameter for ONLY those 4 has the flag set to continue on error = true. ALL the other copies are continue on error=false.

It looks like those copy commands and continue on error= true settings have not changed for about 2 years. Somebody recognized it needed to be able to continue and that those would have an error but still needed to be there.

Can we have that be a separate task/PR  to investigate about why those copies are like that? I'm guessing that if the continue on error was set to true in the past there was a decent reason for it but who knows.

Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj]: Nice catch. I didn't have display of spaces/tabs turned on and didn't see that what I pasted from an external source had the tabs.
My setup doesn't add tabs just the paste from outside.

Tabs removed now.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj [hunk 7]
```diff
@@ -11,6 +11,12 @@
...
+	    $(MSBuildWarningsAsMessages);MSB3277;MSB3030
+    </MSBuildWarningsAsMessages>
+  </PropertyGroup>
+
```
[the remark this answers is not in the document]
Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj]: Tabs removed

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 3 above)
Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: Michael wanted me to suppress the Newtonsoft conflict warnings.

As for the 4 files copying issue (that looks like it has been there for a couple of years). All this does is suppress the message that on at least one of the copies it fails (which is why the continue on error=true is there and only on those 4 copies).
This just suppresses the warning but doesn't affect the proper copying of the files.

File: Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs (hunk 5 above)
Comment (schlaman-ms) [reply Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs]: Ok now there is better XML comment

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 2 above)
Comment (shaopeng-gh) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:353]: properties not before fields. Constructor not before fields. public not before static. --- got it.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 3 above)
Comment (michaelcfanning) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: When you disable warnings in project files, how about adding an XML comment describing what's being suppressed?

File: Src/Plugins/Security/Security.csproj [hunk 8]
```diff
@@ -11,6 +11,12 @@
     <PackageId>Sarif.PatternMatcher.Security</PackageId>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <MSBuildWarningsAsMessages>
+        $(MSBuildWarningsAsMessages);MSB3030
```
Comment (michaelcfanning) [inline Src/Plugins/Security/Security.csproj:16]: >MSB3030 [](http://example.com/codeflow?start=37&length=7)

Add comment here and elsewhere documenting what this warning is.

File: Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs [hunk 9]
```diff
@@ -39,6 +39,9 @@ public AlibabaEcsRequestSigner(DateTime? dateTime = null, string signatureNonce
         /// Sign the HTTP request using Alibaba Cloud access key and secret so it
         /// can be used to access Alibaba Cloud service REST API.
         /// </summary>
+        /// <param name="request">Unsigned httpRequest passed in that will be signed for Alibaba Cloud service REST API.</param>
```
Comment (michaelcfanning) [inline Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs:42]: >httpRequest [](http://example.com/codeflow?start=43&length=11)

what is this term? it's not a variable name, it's not a type, it's not a valid english term. ? :)

File: Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs [hunk 10]
```diff
@@ -39,6 +39,9 @@ public AlibabaEcsRequestSigner(DateTime? dateTime = null, string signatureNonce
...
+        /// <param name="accessKey">Alibbaba Cloud access key.</param>
```
Comment (michaelcfanning) [inline Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs:43]: >Alibbaba [](http://example.com/codeflow?start=36&length=8)

this is misspelled.

how much of an investment are you making in improving the doc comments? If we aren't going to put them in a sufficient level of polish, we shouldn't bother.

Also, weren't you going to broadcast the engineering standards to the group before committing to many fixes? i.e., was there a discussion with the team about our position on this code standard? If we don't like it, we can just disable the check. Please do schedule this conversation, probably invite the whole team and you can review individual checks with us.

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs [hunk 11]
```diff
@@ -76,9 +76,10 @@ public AnalyzeCommand(IFileSystem fileSystem = null)
                     FileVersionInfo fvi = fileSystem.FileVersionInfoGetVersionInfo(Path.Combine(directory, definitions.ValidatorsAssemblyName));
 
                     name = $"{fvi?.CompanyName}/{fvi?.FileDescription}/{name}";
+
                     // TBD add version details. Breaks test baselines currently.
-                    //semanticVersion = fvi?.ProductVersion;
-                    //version = fvi?.FileVersion;
+                    // semanticVersion = fvi?.ProductVersion;
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher/AnalyzeCommand.cs:81]: >semanticVersion [](http://example.com/codeflow?start=23&length=15)

So here we've made a code clean-up while ignoring  something important, a dev (me! eep) has left a work item undone. Far more important than injecting spaces is that devs look at this and either update the code comment or create a tracking issue for the work.

Welcome to code clean-up. :) This will rub your face in all our technical debt. We perform code clean-up not merely to make symptoms (warnings) for away, we do it to rub our faces in our debt and potentially drive clean-up.

Please file an issue for this open item, which is actually pretty important TODO.

File: Src/Sarif.PatternMatcher/AnalyzeContext.cs [hunk 12]
```diff
@@ -13,6 +13,16 @@ public class AnalyzeContext : AnalyzeContextBase
     {
         public static long FilesFilteredBySniffRegex;
 
+        /// <summary>
+        /// Gets or sets a dictionary linking file text with
+        /// A String8 that is used to in RE2 searching.
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher/AnalyzeContext.cs:18]: this isn't grammatically correct.

File: Src/Sarif.PatternMatcher/AnalyzeContext.cs [hunk 13]
```diff
@@ -13,6 +13,16 @@ public class AnalyzeContext : AnalyzeContextBase
...
+        /// An array of bytes that comprise a buffer used in String8 conversion
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher/AnalyzeContext.cs:19]: >array [](http://example.com/codeflow?start=15&length=5)

I think the intent was to encode some of this text as XML params args, but that wasn't done.

Comment (michaelcfanning) [reply Src/Sarif.PatternMatcher/AnalyzeContext.cs:19]: I now understand that you simply moved this code. If you touched it, you bought it. :) The comment is bad.

Also, please do review this code standard with the team, i.e., property placement. Are we really going to enforce that standard? Why exactly? Someone on the team can advocate for it, I would like to hear the rationale.

Notice in this file as well that you moved a property with a code comment, but we're ignoring every other public thing that has no code comments. No good.

Now, balancing my remarks, if we get to a clean condition, we can enable warnings as errors and help prevent regression. So that is a fine value that might argue for not getting lost in too many 'you touched it' morasses.

File: Src/Sarif.PatternMatcher/JsonLogicalLocationProcessor.cs [hunk 14]
```diff
@@ -27,6 +67,9 @@ public class JsonLogicalLocationProcessor
         ///     - For arrays, one might report the whole array, but this gets "closer" to the match.
         ///     - For objects, spanning properties will get "closer" to the match.
         /// </remarks>
+        /// <param name="results">collection of JSON results to process.</param>
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher/JsonLogicalLocationProcessor.cs:70]: >collection [](http://example.com/codeflow?start=34&length=10)

I don't think that most doc comments of this kind are sentence fragments. Please review actual .NET docs comments to get a sense of what the standard is.

File: Src/Sarif.PatternMatcher.Sdk/Crc32.cs [hunk 15]
```diff
@@ -7,6 +7,10 @@ namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk
 {
     public static class Crc32
     {
+        public static readonly uint[] Crc32IEEETable = CreateCrcTable(Crc32IEEEPolynomial);
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher.Sdk/Crc32.cs:10]: >public [](http://example.com/codeflow?start=8&length=6)

Why is this data public?

File: Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs (hunk 6 above)
Comment (michaelcfanning) [reply Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs:36]: S_ is not a convention for any standard that I'm aware of.

:) if we prefer s_ it needs to remain that way and you'll need to disable warnings.

File: Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs [hunk 16]
```diff
@@ -29,11 +29,11 @@ internal class SpamTestRule : Skimmer<AnalyzeContext>, IOptionsProvider
         internal static Random s_random = new Random(s_seed);
 
         [ThreadStatic]
-        internal static TestRuleBehaviors s_testRuleBehaviors;
+        private static TestRuleBehaviors s_testRuleBehaviors;
```
Comment (michaelcfanning) [inline Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs:32]: >private [](http://example.com/codeflow?start=8&length=7)

You are going to break upstream consumers of this property, I believe. We shouldn't do it but this data is visible to some SPAM things, I think, check the internals visible to attribute data.

The 'right' thing to do would be to get over it and just elevate this to a public static property.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 3 above)
Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: Added very complete XML comments to explain the 2 suppressions.

File: Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs (hunk 10 above)
Comment (schlaman-ms) [reply Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs]: Fixed the embarrassingly bad misspell of Alibaba (oddly I chose to only misspell in one location - did the key get stuck down ??) :)

Setup a meeting to discuss coding standards in detail.

File: Src/Sarif.PatternMatcher.Sdk/Crc32.cs (hunk 15 above)
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher.Sdk/Crc32.cs]: The existing code had it as public but it should have been private as that field is not referenced outside of this class and clearly should be private.
Changed it to private.

File: Src/Sarif.PatternMatcher.Sdk/ExtensionMethods.cs [hunk 17]
```diff
@@ -10,6 +10,8 @@ namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk
 {
     public static class ExtensionMethods
     {
+        public static char RedactionChar => '?';
```
[the remark this answers is not in the document]
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher.Sdk/ExtensionMethods.cs:13]: Yes it is and I have changed it to a const char.

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 11 above)
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:81]: New user story created:

[Add version details to ToolComponent in CreateSkimmers](https://dev.azure.com/mseng/1ES/_workitems/edit/2066916)

Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:81]: There are about 50 TODOs and about 13 TBDs in BinSkim.

File: Src/Sarif.PatternMatcher/AnalyzeContext.cs (hunk 13 above)
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher/AnalyzeContext.cs:19]: I do have to say the 'you touched it' idea can be a tad onerous and discourage doing a lot of cleanup that has high value but is too tedious to attack the owned content in the same pass.

I think there is great value in having multiple passes/waves of cleanup with different types of things done in each wave and if you mix too many things it makes it quite inefficient.

And personally, I would be very happy to NOT advocate for caring about code or property placement in a file.
It does however, feel messy to have fields/properties after the construction and methods. It does make sense to have private methods at the end of the class but that is a minor concern. There are so many subtle "this before this" rules it can become an almost unsolvable puzzle to find the perfect combination that satisfies all the placement rules.

File: Src/Sarif.PatternMatcher/AnalyzeContext.cs (hunk 12 above)
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher/AnalyzeContext.cs:18]: This falls under the should the 'if you touch it, you own it' rule.

Ultimately if we decide we don't care about the order then this would not have come up and there are literally hundreds of comments I'm guessing that need help and that takes away time from perhaps other higher yield tasks or  perhaps a dedicated cleanup pass for comments.

File: Src/Sarif.PatternMatcher/JsonLogicalLocationProcessor.cs (hunk 14 above)
Comment (schlaman-ms) [reply Src/Sarif.PatternMatcher/JsonLogicalLocationProcessor.cs]: Improved the XML comments.
Aware that typical comments are more complete.
Had to stop to study code to get a sense of something more appropriate to write.
Trying to balance where to spend time.

File: Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs (hunk 16 above)
Comment (schlaman-ms) [reply Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs]: Ok, now I remember, the VS auto fix was to create a property and s_testRuleBehaviors turned into the backing field which is why it could become private but the property became S_testRuleBehaviors.

But, I have restored everything back and for now there is a warning about it:
warning CS0649: Field 'SpamTestRule.s_testRuleBehaviors' is never assigned to, and will always have its default value

File: Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs (hunk 6 above)
Comment (schlaman-ms) [reply Src/Test.UnitTests.Sarif.PatternMatcher/SpamTestRule.cs:36]: Addressed this in the prior comment about s_testRuleBehaviors.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj [hunk 18]
```diff
@@ -11,6 +11,21 @@
     <PackageId>Sarif.PatternMatcher.AzureDevOpsConfiguration</PackageId>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <MSBuildWarningsAsMessages>
+		// MSB3277 - Found conflicts with different versions of the same
```
Comment (suvamM) [inline Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:16]: Nit: `.` at the end of the sentence. Also "found conflicts *among* different versions..." might be better.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj [hunk 19]
```diff
@@ -11,6 +11,21 @@
...
+		//  dependent assembly that could not be resolved. Specifically
+		//  for Newtonsoft where we need older versions for compatibility
+		// MSB3030 - This error occurs when a file that was supposed to be
+		//  copied isn't found at the expected location.
+		//  There are copy commands that have a continueOnError=true because
```
Comment (suvamM) [inline Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:21]: The text seems a little convoluted to me. How about: "Some copy commands have continueOnError=true since the copy fails, and then eventually succeeds. MSB3030 errors, in such cases, are false positives and are therefore OK to suppress. 
TODO: Investigate why the copy failure occurs even though the operation eventually succeeds."

I am not sure what value is added by saying this has been the case for almost two years. Better to add the TODO so that we can track this, and hopefully mitigate it properly.

File: Src/Plugins/SalModernization/SalModernization.csproj (hunk 4 above)
Comment (suvamM) [reply Src/Plugins/SalModernization/SalModernization.csproj:27]: @schlaman-ms To me, the problems of the duplication is showing up already. If you make a refinement to the XML comment, that adds a follow-on work item to make sure it's updated everywhere. The `build.common.props` was just a suggestion. Can we not create a new suppression file, and make the relevant projects reference that instead? Do we currently have a scenario where we are suppressing things other than MSB3277 and MSB3030?

File: Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs (hunk 9 above)
Comment (suvamM) [reply Src/Plugins/Security/Utilities/AlibabaEcsRequestSigner.cs:42]: +1. I couldn't parse this.

File: Src/Sarif.PatternMatcher/AnalyzeCommand.cs (hunk 11 above)
Comment (suvamM) [reply Src/Sarif.PatternMatcher/AnalyzeCommand.cs:81]: @schlaman-ms Thanks for opening the work item. But I think there is a subtlety here. To me, the value of work items come from the fact that it forms a connected component. For the work item you created, there are the following issues:

1. If someone is browsing the code and reaches this TODO, there is no way to click and jump to the corresponding work item.
2. If someone is browsing our ADO board and reaches the work item, there is no way to point back to this code.
3. Since this is open-source, and we drop in an ADO work-item link, it will be broken for most people. 

We should think of a coding standard for this as well:

1. For Todos in open-source, raise issues. Ensure the issue is linked from code, and there are links to the code from the issue.
2. For internal projects, replace "issues" with ADO work items.

@michaelcfanning what do you think about this suggestion?

File: Src/Sarif.PatternMatcher/AnalyzeContext.cs (hunk 13 above)
Comment (suvamM) [reply Src/Sarif.PatternMatcher/AnalyzeContext.cs:19]: I have a preference for adding public methods before private ones: if someone else is reading the class to understand how to _interact_ with it (and not necessarily how it implements things internally), then having the public stuff at the top makes it easier to understand. One doesn't need to traverse different parts of the class. 

Personal preference here.

**Review by suvamM — CHANGES_REQUESTED** [review]

File: Src/Sarif.PatternMatcher.Sdk/Crc32.cs [hunk 20]
```diff
@@ -7,6 +7,9 @@ namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk
 {
     public static class Crc32
     {
+        public static readonly uint[] Crc32CastagnoliTable = CreateCrcTable(Crc32CastagnoliPolynomial);
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher.Sdk/Crc32.cs:10]: >public [](http://example.com/codeflow?start=8&length=6)

Why is this public? This public array of items is mutable, i.e., any caller can change index 0 this data and break it. That's not good.

This data should either not be public or we should be exposing an immutable array.

File: Src/Sarif.PatternMatcher/AnalyzeContext.cs (hunk 12 above)
Comment (michaelcfanning) [reply Src/Sarif.PatternMatcher/AnalyzeContext.cs:18]: Yes, you touched it. When you touch code, in our project, people should take a careful look at it. 

Here is an updated code comment. I invested more time in discussing this with you than authoring it. And btw - improving code comments is a high yield task in my opinion. :)


    /// <summary>
   /// Gets or sets a map that associates file text with data used in RE2
   /// searching, specifically, a String8 form of the file text and arrays
   /// that comprise a map of UTF8 to UTF16 byte indices. This data is
   /// required to rationalize RE2 match segments with original .NET strings.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 3 above)
Comment (michaelcfanning) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: So, this is an appeal to authority, and I encourage everyone to reject comments on the basis of 'Michael said so'. Instead, please try to communicate the technical reason I provided: the SARIF SDK is consumed by other projects that themselves may consume outdated versions of Newtonsoft. If we update our references, we will force those projects to upgrade as well, which may invalidate our assembly from reuse. 

I say lots of things, and I'm often wrong or my information is incomplete. Trust me if you haven't had direct experience of it already.

Comment (schlaman-ms) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: So a better answer would be:

We have consumers of this tool that use older frameworks and if we updated to newer versions of Newtonsoft then that would break the compatibility with consumers that are not as up to date.
Because of that, it is necessary to maintain older versions within the tool and this creates warnings that there are version conflicts. For this reason I suppressed the warnings rather than update the versions.

File: Src/Plugins/SalModernization/SalModernization.csproj (hunk 4 above)
Comment (shaopeng-gh) [reply Src/Plugins/SalModernization/SalModernization.csproj:27]: I agree with it is better if possible, to define in one place, even if it applies to more projects those not have to have it.
applies to more projects has not have major impact.
The benefit is 
less duplicate code, 
easier to maintain, 
and if the project that currently does not need starts to have the e.g. newton in the future then it is already covered no need to add one by one again.

File: Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj (hunk 3 above)
Comment (shaopeng-gh) [reply Src/Plugins/AzureDevOpsConfiguration/AzureDevOpsConfiguration.csproj:15]: I think this is wrong, the correct xml comment should be: `<!-- ... -->` ?
